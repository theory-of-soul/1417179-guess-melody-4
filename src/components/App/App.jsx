import React from "react";
import PropTypes from 'prop-types';
import WelcomeScreen from "../WelcomeScreen/WelcomeScreen";
import ArtistQuestionScreen from "../ArtistQuestionScreen/ArtistQuestionScreen";
import {BrowserRouter, Route, Switch} from "react-router-dom";
import GenreQuestionScreen from "../GenreQuestionScreen/GenreQuestionScreen";
import withAudioPlayer from "../../HOC/withAudioPlayer/withAudioPlayer";
import GameScreen from "../GameScreen/GameScreen";
import {connect} from "react-redux";
import {GameType} from "./GameType";
import withMultiSelectAnswers from "../../HOC/withMultiSelectAnswers/withMultiSelectAnswers";
import FailScreen from "../FailScreen/FailScreen";
import WinScreen from "../WinScreen/WinScreen";
import {actionCreator} from "../../reducers/game/game";
import {createOperations} from "../../reducers/data/data";

const GenreQuestionScreenWithPlayer = withAudioPlayer(withMultiSelectAnswers(GenreQuestionScreen));
const ArtistQuestionScreenWithPlayer = withAudioPlayer(ArtistQuestionScreen);

class App extends React.PureComponent {
  constructor(props) {
    super(props);
    this._onUserClickAnswer = this._onUserClickAnswer.bind(this);
    this._getGameScreen = this._getGameScreen.bind(this);
    this._onWelcomeButtonClick = this._onWelcomeButtonClick.bind(this);
    this._onReplayButtonClick = this._onReplayButtonClick.bind(this);
  }

  componentDidMount() {
    this.props.loadQuestions();
  }

  _onWelcomeButtonClick() {
    this.props.onNextStep();
  }

  _onReplayButtonClick() {
    this.props.onResetGame();
  }

  _onUserClickAnswer(question, userAnswer) {
    const {
      onCheckAnswer,
      onNextStep
    } = this.props;

    onCheckAnswer(question, userAnswer);
    onNextStep();
  }

  _getGameScreen() {
    const {
      step,
      userErrors,
      questions,
      errorAmount
    } = this.props;

    const nextGameQuestion = this.props.questions[step];
    const welcomeScreenStepNumber = -1;

    if (step === welcomeScreenStepNumber) {
      return (
        <WelcomeScreen
          errorAmount={errorAmount}
          onClickHandler={this._onWelcomeButtonClick}
        />
      );
    }

    if (userErrors >= errorAmount) {
      return (
        <FailScreen onClickReplayHandler={this._onReplayButtonClick}/>
      );
    }

    if (!nextGameQuestion) {
      return (
        <WinScreen
          questionAmount={questions.length}
          errorAmount={userErrors}
          onClickReplayHandler={this._onReplayButtonClick}
        />
      );
    }

    if (nextGameQuestion && nextGameQuestion.type) {
      if (nextGameQuestion.type === GameType.ARTIST) {
        return (
          <GameScreen userErrors={userErrors}>
            <ArtistQuestionScreenWithPlayer
              question={questions[step]}
              handleAnswer={this._onUserClickAnswer}
            />
          </GameScreen>
        );
      }
      if (nextGameQuestion.type === GameType.GENRE) {
        return (
          <GameScreen userErrors={userErrors}>
            <GenreQuestionScreenWithPlayer
              question={questions[step]}
              handleAnswer={this._onUserClickAnswer}
            />
          </GameScreen>
        );
      }
    }

    return null;
  }

  render() {
    const {
      questions
    } = this.props;

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">
            {this._getGameScreen()}
          </Route>
          <Route exact path="/dev-artist">
            <GameScreen userErrors={1}>
              <ArtistQuestionScreenWithPlayer
                question={questions[0]}
                handleAnswer={this._onUserClickAnswer}
              />
            </GameScreen>
          </Route>
          <Route exact path="/dev-genre">
            <GameScreen userErrors={2}>
              <GenreQuestionScreenWithPlayer
                question={questions[1]}
                handleAnswer={this._onUserClickAnswer}
              />
            </GameScreen>
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

const artistQuestionType = PropTypes.shape({
  type: PropTypes.string.isRequired,
  rightAnswer: PropTypes.string.isRequired,
  answers: PropTypes.arrayOf(
      PropTypes.shape({
        pic: PropTypes.string.isRequired,
        name: PropTypes.string.isRequired,
      })
  ).isRequired
});

const genreQuestionType = PropTypes.shape({
  type: PropTypes.string.isRequired,
  genre: PropTypes.string.isRequired,
  answers: PropTypes.arrayOf(
      PropTypes.shape({
        genre: PropTypes.string.isRequired,
        audioSrc: PropTypes.string.isRequired
      })
  ).isRequired
});

App.propTypes = {
  errorAmount: PropTypes.number.isRequired,
  questions: PropTypes.arrayOf(PropTypes.oneOfType([artistQuestionType, genreQuestionType])).isRequired,
  step: PropTypes.number.isRequired,
  onNextStep: PropTypes.func.isRequired,
  onCheckAnswer: PropTypes.func.isRequired,
  onResetGame: PropTypes.func.isRequired,
  userErrors: PropTypes.number.isRequired,
  loadQuestions: PropTypes.func.isRequired
};

const mapStateToProps = (state) => {
  return {
    step: state.step,
    userErrors: state.errors,
    questions: state.questions,
    errorAmount: state.maxErrors
  };
};

const mapDispatchToProps = (dispatch) => {
  return {
    onNextStep: () => {
      dispatch(actionCreator.nextStep());
    },
    onCheckAnswer: (question, userAnswer) => {
      dispatch(actionCreator.increaseErrors(question, userAnswer));
    },
    onResetGame: () => {
      dispatch(actionCreator.resetGame());
    },
    loadQuestions: () => {
      dispatch(createOperations.loadQuestions());
    }
  };
};

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
