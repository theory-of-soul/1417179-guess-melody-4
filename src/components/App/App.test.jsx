import React from 'react';
import renderer from 'react-test-renderer';
import {App} from './App';

const errorAmounts = 3;
const questions = [{
  type: `artist`,
  audioSrc: ``,
  rightAnswer: `Пелагея`,
  answers: [{
    pic: `http://placehold.it/134x134`,
    name: `Пелагея`
  }, {
    pic: `http://placehold.it/134x134`,
    name: `Краснознаменная дивизия имени моей бабушки`
  }, {
    pic: `http://placehold.it/134x134`,
    name: `Lorde`
  }]
}, {
  type: `genre`,
  genre: `инди-рок`,
  answers: [{
    genre: `инди-рок`,
    audioSrc: ``,
  }, {
    genre: `рок`,
    audioSrc: ``,
  }, {
    genre: `поп`,
    audioSrc: ``,
  }, {
    genre: `рок`,
    audioSrc: ``,
  }]
}];

describe(`App snapshot tests`, () => {
  it(`App should show WelcomeScreen`, () => {
    const component = renderer
      .create(
          <App
            errorAmount={errorAmounts}
            questions={questions}
            step={-1}
            onNextStep={() => {}}
            onCheckAnswer={() => {}}
            userErrors={0}
            onResetGame={() => {}}
            hasError={false}
            loadQuestions={() => {}}
            userAlreadyAuth={false}
            loginUser={() => {}}
          />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });

  it(`App should show Auth screen`, () => {
    const component = renderer
      .create(
          <App
            errorAmount={errorAmounts}
            questions={questions}
            step={questions.length}
            onNextStep={() => {}}
            onCheckAnswer={() => {}}
            userErrors={1}
            onResetGame={() => {}}
            loadQuestions={() => {}}
            hasError={false}
            userAlreadyAuth={false}
            loginUser={() => {}}
          />
      )
      .toJSON();

    expect(component).toMatchSnapshot();
  });
});
