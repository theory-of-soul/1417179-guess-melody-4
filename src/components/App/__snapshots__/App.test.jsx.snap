// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App snapshot tests App should show Auth screen 1`] = `
<section
  className="login"
>
  <div
    className="login__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <h2
    className="login__title"
  >
    Вы настоящий меломан!
  </h2>
  <p
    className="login__text"
  >
    Хотите узнать свой результат? Представьтесь!
  </p>
  <form
    action=""
    className="login__form"
    onSubmit={[Function]}
  >
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="name"
      >
        Логин
      </label>
      <input
        className="login__input"
        id="name"
        name="name"
        type="text"
      />
    </p>
    <p
      className="login__field"
    >
      <label
        className="login__label"
        htmlFor="password"
      >
        Пароль
      </label>
      <input
        className="login__input"
        id="password"
        name="password"
        type="password"
      />
      <span
        className="login__error"
      >
        Неверный пароль
      </span>
    </p>
    <button
      className="login__button button"
      type="submit"
    >
      Войти
    </button>
  </form>
  <button
    className="replay"
    onClick={[Function]}
    type="button"
  >
    Сыграть ещё раз
  </button>
</section>
`;

exports[`App snapshot tests App should show WelcomeScreen 1`] = `
<section
  className="welcome"
>
  <div
    className="welcome__logo"
  >
    <img
      alt="Угадай мелодию"
      height="83"
      src="img/melody-logo.png"
      width="186"
    />
  </div>
  <button
    className="welcome__button"
    onClick={[Function]}
  >
    <span
      className="visually-hidden"
    >
      Начать игру
    </span>
  </button>
  <h2
    className="welcome__rules-title"
  >
    Правила игры
  </h2>
  <p
    className="welcome__text"
  >
    Правила просты:
  </p>
  <ul
    className="welcome__rules-list"
  >
    <li>
      Нужно ответить на все вопросы.
    </li>
    <li>
      Можно допустить 
      3
       ошибки.
    </li>
  </ul>
  <p
    className="welcome__text"
  >
    Удачи!
  </p>
</section>
`;
